Python3 support

utzig opened the pull request.

utzig committed:
print as funtion

utzig committed:
Fix StringIO import

utzig committed:
Use encode_hex/decode_hex, bytestrings

utzig committed:
Literal strings as bytes

utzig committed:
Update encode_hex/decode_hex, strings as bytes

utzig committed:
Use integer division, encode_hex, strings as bytes

utzig committed:
Use range instead of xrange

utzig committed:
Use encode_hex/decode_hex, strings as bytes

utzig committed:
Absolute imports

utzig committed:
Absolute imports + decode_hex

utzig committed:
Update for safe_ord/ascii_chr/str_to_bytes/decode_hex

utzig committed:
Use strings as bytes, integer division, is_integer

utzig committed:
Use tuples to index dict keys

utzig committed:
Use strings as bytes, encode_hex, is_integer

utzig committed:
Enable nodes to be hashed in Py3

utzig committed:
Absolute imports

utzig committed:
Use safe_ord/ascii_chr, strings as bytes

utzig committed:
Absolute imports

utzig committed:
Absolute imports

utzig committed:
Use dict values

utzig committed:
Absolute imports, decode_hex/safe_ord, strings as bytes, integer division

utzig committed:
unicode in Py2, str in Py3

utzig committed:
Absolute imports, decode_hex, use as in exceptions

pipermerriam commented on devp2p/discovery.py:
```diff
@@ -218,7 +222,7 @@ def __init__(self, app, transport):
         "return node or create new, update address if supplied"
-        assert isinstance(nodeid, str)
+        assert isinstance(nodeid, bytes)
         assert len(nodeid) == 512 / 8
```
Just noticed this isn't using floored division `//`

utzig replied:
Yep, I did not change it because the result will always be a whole number (where float == int). Will fix it.

pipermerriam commented on devp2p/discovery.py:
```diff
@@ -218,7 +222,7 @@ def __init__(self, app, transport):
     def get_node(self, nodeid, address=None):
         "return node or create new, update address if supplied"
-        assert isinstance(nodeid, str)
+        assert isinstance(nodeid, bytes)
```
This statement isn't exactly equivalent in 2/3 since it will return `True` for both byte and text strings in python 2 but only true for byte strings in python3 .

utzig replied:
Since `str == bytes` in Python 2, I guess this is want we want. What error condition do you have in mind?

pipermerriam replied:
Nevermind.  This isn't a problem.  I shouldn't review pull requests prior to having my morning ☕

utzig commented on devp2p/tests/test_discovery.py:
```diff
@@ -128,15 +129,15 @@ def test_ping_pong():
     4be2be5a685a80971ddcfa80cb422cdd0101ec04cb847f000001820cfa8215a8d790000000000000
     000000000000000000018208ae820d058443b9a3550102
-    '''.translate(None, ' \t\n').decode('hex'),
+    '''.replace(b' ', b'').replace(b'\n', b'')),
```
I don't like these two `replace` calls here. Done it because `translate` works slight differently between Py2/Py3. Maybe creating an utility function like `remove_chars_from_str` or similar would be better?

janx replied:
+1

utzig committed:
Use floor division

utzig committed:
Fix tabulation in string formatting

utzig committed:
Remove unneeded section header

utzig committed:
Fix extra relative imports

utzig committed:
Remove more relative imports

utzig committed:
Encode/decode fixes

utzig committed:
More strings as bytes conversions

janx commented on devp2p/rlpxcipher.py:
```diff
@@ -94,7 +97,9 @@ def decrypt_header(self, data):
-        def mac(data=''):
+        def mac(data=b''):
+            if not isinstance(data, bytes):
+                data = bytes(data, 'utf-8')
```
`str_to_bytes` can be used here.

janx commented:
`ascii_chr` is missing from `rlp.utils_py2`.

utzig committed:
Use str_to_bytes

utzig commented:
@janx I see it here: https://github.com/ethereum/pyrlp/blob/develop/rlp/utils_py2.py#L17. Am I missing something?

utzig committed:
Mark full app tests with xfail for Python3.

The error that is happening is very similar to the one that happens
with PyPy already reported and can be investigated together.
See ethereum/pydevp2p#37.

janx commented:
Aha I overlooked those aliases, again.

@utzig :+1:

utzig committed:
Fix Python 3.4.x bytes string formatting

utzig committed:
Skip known py3/pypy incompatible tests instead of xfailing

utzig committed:
Remove xfail/skip from tests that seem to run fine

utzig committed:
Fix integer assert

utzig committed:
Remove skipif from good test

utzig committed:
Add remove_chars, flake8 fixes

utzig committed:
Use remove_chars func, more bytes strings and flake8 fixes

utzig committed:
Only skip full app tests for python3/pypy3

utzig committed:
More py3/py2 compatibility fixes

utzig committed:
Switch skipif back to xfail

utzig committed:
More py2/py3 compatibility fixes

utzig committed:
Avoid deleting dict elements while iterating in Py3

utzig committed:
More Py3 fixes, remove xfail marks for Py3

utzig committed:
Use more recent pypy2

utzig commented:
@janx @pipermerriam Guys, from my perspective this is completed. Is there anything that you are still expecting? Any other suggestions?

pipermerriam commented on devp2p/discovery.py:
```diff
@@ -62,7 +66,7 @@ def __init__(self, ip, udp_port, tcp_port=0, from_binary=False):
             # see https://docs.python.org/2/library/socket.html#socket.getaddrinfo
             ips = [
-                unicode(ai[4][0])
+                str(ai[4][0]) if PY3 else unicode(ai[4][0])
```
The use of `unicode` here bugs me.  I'm not sure it's exactly a problem, but `unicode` isn't defined in python 3 so it's weird to have it inlined even though it won't execute in the py3 environment.

This is more of a nitpick but this would be cleaner to just have a `force_text` function and remove the `if/else`.  Again, just a nitpick.  This is totally functional as is.

utzig replied:
I didn't remember why this was required but now checking again, it seems that `ipaddress.ip_address` only accepts `unicode`!

https://github.com/phihag/ipaddress/blob/master/ipaddress.py#L159

utzig replied:
Btw, since most of our helpers come from `rlp.utils`, the right place to add an `force_text`, `force_unicode` or similar would be there, but I would prefer to do it later.

pipermerriam commented on devp2p/kademlia.py:
```diff
@@ -480,17 +484,17 @@ def _expected_pongs():
-        # check and removed timedout find requests
-        for nodeid, timeout in self._find_requests.items():
+        # check and removed timed out find requests
+        for nodeid, timeout in tuple(self._find_requests.items()):
```
Another nitpick.  

I dislike that this dictionary is modified in place.  Mutability leads to terrible things and whenever possible I always try to avoid it.

It could be replaced with a comprehension to replace it with an entirely new dictionary which would remove part the mutation aspect.

``` python
self._find_requests = {
    nodeid: timeout
    for nodeid, timeout in self._find_requests.items()
    if time.time() <= timeout
}
```

pipermerriam commented on devp2p/kademlia.py:
```diff
@@ -421,22 +425,22 @@ def _expected_pongs():
 
         # check for timed out pings and eventually evict them
-        for _pingid, (timeout, _node, replacement) in self._expected_pongs.items():
+        for _pingid, (timeout, _node, replacement) in list(self._expected_pongs.items()):
```
Same comment here about the mutation aspect.  I'd replace this with the following code.

``` python
# Gather the id's and nodes that need to be removed.
pind_ids_to_delete = {
    _pingid: _node
    for _pingid, (timeout, _node, replacement)
    in self._expected_pongs.items()
    if time.time() > timeout
}
# replace the tracking dictionary with an updated one with the timeoud out items removed.
self._expected_pongs = {
    _pingid: (timeout, _node, replacement)
    for _pingid, (timeout, _node, replacement)
    in self._expected_pongs.items()
    if _pingid not in ping_ids_to_delete
}
# Log and execute the removals
for _pingid, _node in ping_ids_to_delete.items():
    self._deleted_pingids.add(_pingid)
    self.routing.remove_node(_node)
```

utzig replied:
In relation to the mutation I don't see a real problem in this specific case, because we're creating a list of the items to iterate over while deleting elements in original dict. Also the logging inside might be useful for tracking issues.

pipermerriam commented on devp2p/rlpxcipher.py:
```diff
@@ -114,7 +118,9 @@ def decrypt_body(self, data, body_size):
-        def mac(data=''):
+        def mac(data=b''):
+            if not isinstance(data, bytes):
+                data = bytes(data, 'utf-8')
```
Same here with consistent use of `str_to_bytes`

pipermerriam commented on devp2p/tests/test_discovery.py:
```diff
@@ -272,7 +277,7 @@ def test_bootstrap_udp():
         num = len(kproto(app).routing)
-        print num
+        print(num)
         if i < len(apps) / 2:  # only the first half has enough time to get all updates
```
Dunno if this should be replaced with `//` for floored division.  May not matter but this will cause the tests run differently in py2/py3 if `num` is odd.

pipermerriam commented on devp2p/tests/test_peermanager.py:
```diff
@@ -35,7 +36,7 @@ def test_app_restart():
 
     # Start the app 10-times: there should be no exception like 'Bind error'
-    for i in xrange(0, 10):
+    for i in range(0, 10):
```
nitpick.

This is the same as `for i in range(10)`

pipermerriam commented on devp2p/tests/test_slogging.py:
```diff
@@ -1,7 +1,10 @@
+try:
+    from StringIO import StringIO
+except ImportError:
+    from io import StringIO
```
nitpick

You this line works for both py2 and py3 so you can remove the try/catch and replace it with this single import.

utzig replied:
So I tried leaving only an `from io import StringIO` which failed the previous build. The reason is that `StringIO.StringIO` accepts both bytes and unicode while `io.StringIO` accepts only unicode.

https://mail.python.org/pipermail/python-list/2013-May/648245.html

I think it makes more sense to try `StringIO.StringIO` than going through all the log calls in the project to change string literals.

pipermerriam commented:
@utzig looks good.  I left a bunch of comments but I think all of them were just nitpick/stylistic things that don't need to stand in the way of getting this merged.

utzig commented:
@pipermerriam Thanks a lot, that was thorough! I will check the issues you found as soon as I get some free time.

utzig committed:
Merge remote-tracking branch 'upstream/develop' into python3-support

utzig committed:
Fix small review issues

utzig committed:
Use str to bytes

utzig committed:
Use dict comprehension for element removal

utzig committed:
Try using String.String because it acepts bytes

utzig commented:
@pipermerriam Hey! Is there anything still blocking this from being merged?

pipermerriam commented:
Was good to go last time I looked it over. I can take another look later
today to be sure.

On Tue, Oct 11, 2016, 6:25 AM Fabio Utzig notifications@github.com wrote:

> @pipermerriam https://github.com/pipermerriam Hey! Is there anything
> still blocking this from being merged?
> 
> —
> You are receiving this because you were mentioned.
> 
> Reply to this email directly, view it on GitHub
> https://github.com/ethereum/pydevp2p/pull/46#issuecomment-252902445, or mute
> the thread
> https://github.com/notifications/unsubscribe-auth/AAyTgmDfuDAHkTiCKCRVFllwwL5wEcdtks5qy4BMgaJpZM4JJS4F
> .
